File: packages/react-native/android/src/main/java/com/nitropush/HybridLocalPackage.kt
package com.nitropush

import com.margelo.nitro.core.Promise
import com.margelo.nitro.nitropush.HybridLocalPackageSpec
import com.margelo.nitro.nitropush.InstallMode
import com.nitropush.sdk.NitroPushSdk
import com.nitropush.sdk.NPInstallMode
import com.nitropush.sdk.NPLocalPackage

/**
 * Kotlin counterpart of `ios/HybridLocalPackage.swift`. HybridObject wrapping
 * a single [NPLocalPackage] snapshot. Returned from
 * [HybridRemotePackage.download] and the inspection methods on
 * [HybridNitroPushClient] (`getCurrentPackage`, `getPendingPackage`).
 */
internal class HybridLocalPackage(
  private val plain: NPLocalPackage,
) : HybridLocalPackageSpec() {

  // ── Properties ────────────────────────────────────────────────────────────
  override val releaseId: String get() = plain.releaseId
  override val label: String get() = plain.label
  override val packageHash: String get() = plain.packageHash
  override val packageSize: Double get() = plain.packageSize
  override val appVersion: String get() = plain.appVersion
  override val otaVersion: Double? get() = plain.otaVersion
  override val displayVersion: String? get() = plain.displayVersion
  override val platforms: Array<String>? get() = plain.platforms
  override val isMandatory: Boolean get() = plain.isMandatory
  override val description: String? get() = plain.description
  override val isPending: Boolean get() = plain.isPending
  override val isFailedInstall: Boolean get() = plain.isFailedInstall
  override val isFirstRun: Boolean get() = plain.isFirstRun
  override val bundlePath: String get() = plain.bundlePath

  // ── Methods ───────────────────────────────────────────────────────────────
  override fun install(
    installMode: InstallMode,
    minimumBackgroundDuration: Double,
  ): Promise<Unit> = Promise.parallel {
    NitroPushSdk.shared.installUpdate(
      pkg = plain,
      installMode = installMode.toPlain(),
      minimumBackgroundDurationSeconds = minimumBackgroundDuration,
    )
  }

  override fun rollback(): Promise<Unit> = Promise.parallel {
    NitroPushSdk.shared.rollback(plain.releaseId)
  }
}

private fun InstallMode.toPlain(): NPInstallMode = when (this) {
  InstallMode.IMMEDIATE -> NPInstallMode.IMMEDIATE
  InstallMode.ON_NEXT_RESTART -> NPInstallMode.ON_NEXT_RESTART
  InstallMode.ON_NEXT_RESUME -> NPInstallMode.ON_NEXT_RESUME
  InstallMode.ON_NEXT_SUSPEND -> NPInstallMode.ON_NEXT_SUSPEND
}

File: packages/react-native/ios/HybridLocalPackage.swift
//
//  HybridLocalPackage.swift
//  NitroPush
//
//  HybridObject wrapping a single `NPLocalPackage` snapshot. Returned from
//  `RemotePackage.download()` and the inspection methods on the parent
//  singleton (`getCurrentPackage`, `getPendingPackage`).
//

import Foundation
import NitroModules

final class HybridLocalPackage: HybridLocalPackageSpec {
    private let plain: NPLocalPackage

    init(plain: NPLocalPackage) {
        self.plain = plain
        super.init()
    }

    // MARK: - Properties

    var releaseId: String { plain.releaseId }
    var label: String { plain.label }
    var packageHash: String { plain.packageHash }
    var packageSize: Double { plain.packageSize }
    var appVersion: String { plain.appVersion }
    var otaVersion: Double? { plain.otaVersion }
    var displayVersion: String? { plain.displayVersion }
    var platforms: [String]? { plain.platforms }
    var isMandatory: Bool { plain.isMandatory }
    var description: String? { plain.description }
    var isPending: Bool { plain.isPending }
    var isFailedInstall: Bool { plain.isFailedInstall }
    var isFirstRun: Bool { plain.isFirstRun }
    var bundlePath: String { plain.bundlePath }

    // MARK: - Methods

    func install(installMode: InstallMode, minimumBackgroundDuration: Double) throws -> Promise<Void> {
        let snapshot = self.plain
        let mode = installMode.toPlain()
        return Promise.async {
            try await NitroPushSdk.shared.installUpdate(
                pkg: snapshot,
                installMode: mode,
                minimumBackgroundDuration: minimumBackgroundDuration
            )
        }
    }

    func rollback() throws -> Promise<Void> {
        let id = plain.releaseId
        return Promise.async {
            try NitroPushSdk.shared.rollback(releaseId: id)
        }
    }
}

private extension InstallMode {
    func toPlain() -> NPInstallMode {
        switch self {
        case .immediate: return .immediate
        case .onNextRestart: return .onNextRestart
        case .onNextResume: return .onNextResume
        case .onNextSuspend: return .onNextSuspend
        }
    }
}
